dbms/src/Columns/ColumnString.cpp

```diff
@@ -461,75 +460,100 @@ void ColumnString::getPermutationWithCollationImpl(
     }
 }
 
-void ColumnString::updateWeakHash32(
-    WeakHash32 & hash,
-    const TiDB::TiDBCollatorPtr & collator,
-    String & sort_key_container) const
+void updateWeakHash32BinPadding(const std::string_view & view, size_t idx, WeakHash32Info & info)
 {
-    auto s = offsets.size();
+    auto sort_key = BinCollatorSortKey<true>(view.data(), view.size());
+    (*info.hash_data)[idx]
+        = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(sort_key.data), sort_key.size, (*info.hash_data)[idx]);
+}
 
-    if (hash.getData().size() != s)
-        throw Exception(
-            fmt::format(
-                "Size of WeakHash32 does not match size of column: column size is {}, hash size is {}",
-                s,
-                hash.getData().size()),
-            ErrorCodes::LOGICAL_ERROR);
+void updateWeakHash32BinNoPadding(const std::string_view & view, size_t idx, WeakHash32Info & info)
+{
+    auto sort_key = BinCollatorSortKey<false>(view.data(), view.size());
+    (*info.hash_data)[idx]
+        = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(sort_key.data), sort_key.size, (*info.hash_data)[idx]);
+}
 
-    UInt32 * hash_data = hash.getData().data();
+void updateWeakHash32NonBin(const std::string_view & view, size_t idx, WeakHash32Info & info)
+{
+    auto sort_key = info.collator->sortKey(view.data(), view.size(), info.sort_key_container);
+    (*info.hash_data)[idx]
+        = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(sort_key.data), sort_key.size, (*info.hash_data)[idx]);
+}
 
-    if (collator != nullptr)
+void updateWeakHash32NoCollator(const std::string_view & view, size_t idx, WeakHash32Info & info)
+{
+    (*info.hash_data)[idx]
+        = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(view.data()), view.size(), (*info.hash_data)[idx]);
+}
+
+template <typename LoopFunc>
```

Comment:
Seem `LoopFunc` has the fixed type as
```
void LoopFunc(
    const Chars_t & a_data,
    const Offsets & a_offsets,
    WeakHash32Info & info,
    const void()(const std::string_view &, size_t, WeakHash32Info &) & func)
```
, why use template here?

Reply:
changed